Normalize observations with model architecture rather than `VecNormalize`

AdamGleave opened the pull request:
It is often helpful to normalize the observations from an MDP prior to feeding them into a deep policy or reward network. Previously, we used `VecNormalize` to apply this normalization at an environment level. However, this proved to be quite complex for `imitation`. In particular:
  - When serializing policies or rewards, the statistics from `VecNormalize` must be saved, and separately restored. As we do not always have control over the environment wrapping when deserializing, we had to resort to policy and reward wrappers to normalize the observations, significant unnecessary code.
  - When storing buffers of observations, it's important to store unnormalized observations and normalize just before feeding into the policy/reward network to avoid distribution shift. This and related problems were a common source of bugs.

This PR removes any use of `VecNormalize` for normalizing observations -- for now we retain `VecNormalize` to normalize rewards only (with `norm_obs=False` passed in). Instead, we optionally add a layer (for policies, as a feature extractor; for reward networks, as a separate layer) that preprocesses the input to the network in a manner similar to `VecNormalize`. This is largely semantically equivalent (but simplifies the code considerably). The key differences are:

  - Normalization statistics are only updated when the module is in training mode (`self.train()`) and not in evaluation (`self.eval()`). The prior behaviour is hard to characterize exactly -- `VecNormalize` always updated stats (as we never disabled the training mode), but we only sometimes fed data through the `VecNormalize` wrapper.
  - For reward networks, actions (where present) are normalized as well as observations. I think this is generally desirable, but may cause some small differences in results.
  - Some algorithms (e.g. BC, DAgger) didn't previously use `VecNormalize` at all. Because it's now at an architectural level and we're setting normalization by default in the Sacred ingredients `train` and `reward`, these will now have normalization *by default* where previously they did not have it all. It can still be disabled manually though.

I also made a number of other changes to clear up things as I came along them. They're hard to disentangle so keeping them in this PR, but if it's too confusing to review let me know and I'll try to split them up. Key changes:

  - Dropped support for loading `VecNormalize` separately from model. Also dropped support for old-style pickle format for Stable Baselines policies. I don't think we actually have that many artifacts we want to keep around, so it seems worth dropping this backward compatibility code given it's complexity.
  - Various type annotation fixes needed after `pytype` upgrade.
  - Fixes after `gym` upgrade, largely `Pendulum-v0` to `Pendulum-v1` but also some action space tweaks.

Fixes #387

AdamGleave commented:
Normalizating at layer level (`arch_norm`) with `RunningNorm` seems to exactly replicate performance of `VecNormalize` (`vec_norm`), and both do significantly better than no observation normalization in `seals/MountainCar-v0`:

![image](https://user-images.githubusercontent.com/1547683/150697094-bcea47c6-02d2-4a7c-8525-ef168951015a.png)

From:
```bash
parallel --header : --progress --results output/normalize_pr/parallel/ python -m imitation.scripts.train_rl --capture=sys with seals_mountain_car {normalize} seed={seed} common.log_dir=output/normalize_pr/logs/{norm_name}/{seed} ::: seed 0 1 2 ::: normalize "" "normalize_kwargs.norm_obs=False" "normalize_kwargs.norm_obs=False train.normalize_running" :::+ norm_name vec_norm no_norm arch_norm
```

on 9d683d768acd339444299bd3c0ad0511fb143f71

AdamGleave commented:
Normalizing observations at input layer (`arch_norm`) performs comparably to normalizing with `VecNormalize`:

![image](https://user-images.githubusercontent.com/1547683/150701585-fe2a5636-b4a2-491d-b32f-1299d3a990ce.png)

Commit 9f47c1772a2d6376cceaf98a7df8d748c949fdd6 with command:

```bash
parallel --header : --progress --results output/normalize_pr/parallel/ python -m imitation.scripts.train_rl --capture=sys with seals_mountain_car {normal
ize} seed={seed} common.log_dir=output/normalize_pr/logs/{norm_name}/{seed} ::: seed 0 1 2 ::: normalize "" "normalize_kwargs.norm_obs=False" "normalize_kwargs.norm_obs=False train.normalize_running" :::+ norm_name vec_norm no_norm arch_norm
```

[normalize_pr.tar.gz](https://github.com/HumanCompatibleAI/imitation/files/7921642/normalize_pr.tar.gz)

ejnnr commented on src/imitation/scripts/train_preference_comparisons.py:
```diff
@@ -135,13 +125,11 @@ def train_preference_comparisons(
 
-    vec_normalize = None
     reward_net = reward.make_reward_net(venv)
     if reward_net is None:
```
This should never happen now, right? If so, should remove it for clarity

ejnnr reviewed (approved):
I like the new approach to normalization (especially after seeing the implementation, it seems to simplify a couple of annoying things). I didn't spot any major issues, just a couple minor ones that should be easy to fix (and most of which seem optional to me).

Just to check my understanding: this PR already implements the reward model output normalization mentioned in #395, right? So the only thing left to do there is to normalize the input to `SyntheticGatherer`?

AdamGleave commented on src/imitation/algorithms/adversarial/gail.py:
```diff
@@ -67,23 +67,21 @@ def __init__(
-            )
-        self._discriminator = reward_net.to(gen_algo.device)
-        self._reward_net = LogSigmoidRewardNet(self._discriminator)
+        # Raw self._reward_net is discriminator logits
```
Yes, I made the change as we want to pass `reward_net` to parent class. I think possibly I should call it something different in `AdversarialTrainer`, but not sure what. There is just something odd about using `RewardNet` as a discriminator, but everywhere else in `imitation` we use `RewardNet` as a reward network, so I think that terminology is overall better than calling it something more generic like `MDPPredictionNet`.

AdamGleave commented on src/imitation/policies/base.py:
```diff
@@ -65,3 +69,27 @@ class FeedForward32Policy(policies.ActorCriticPolicy):
         super().__init__(*args, **kwargs, net_arch=[32, 32])
+
+
+class NormalizeFeaturesExtractor(torch_layers.FlattenExtractor):
```
Hm, I could just remove the default value and make it more general. I think this is worth doing if we have an application in mind that does something different. Can you think of anything?

ejnnr replied:
Can't come up with another use case after briefly thinking about it, so probably best to leave it as it is

AdamGleave commented:
Thanks for the review, I believe I've addressed most the issues you've highlighted. I'm also dissatisfied with the `self._reward_net` naming convention in `AdversarialTrianer` w.r.t. `GAIL`, but I can't think of anything better, let me know if something comes to mind. 

> Just to check my understanding: this PR already implements the reward model output normalization mentioned in #395, right? So the only thing left to do there is to normalize the input to `SyntheticGatherer`?

I don't think so. We only normalize at the input layer in `networks.build_mlp`. It'd actually be a pretty easy change to also add a normalization at the output layer, so perhaps I should just do that now? But it'd be a bit fiddly ensuring that `VecNormalize` gets removed in the right place, so I lean towards deferring to another PR.

qxcv commented on tests/algorithms/test_adversarial.py:
```diff
@@ -293,7 +303,7 @@ def test_logits_gen_is_high_log_policy_act_prob(
         trans.dones,
     )
-    log_act_prob_non_none = th.as_tensor(np.random.rand(n_timesteps))
+    log_act_prob_non_none = th.as_tensor(np.random.rand(n_timesteps)).to(obs.device)
```
Perhaps the log probs should be negative just for the sake of realism? e.g.

```suggestion
    log_act_prob_non_none = th.as_tensor(np.log(0.1 + 0.9 * np.random.rand(n_timesteps))).to(obs.device)
```

qxcv replied:
(not related to this PR, just seems a bit odd passing in "log probs"  that are not actually log probs)

qxcv commented on tests/policies/test_policies.py:
```diff
@@ -89,18 +80,68 @@ def test_serialize_identity(env_name, model_cfg, normalize, tmpdir):
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        """Returns zeros of same shape as `x`."""
+        assert x.shape[1:] == (self.features_dim,)
+        return x * 0
```
Extremely minor nit: this will produce NaNs if `x` has NaNs or infs. `th.zeros_like` will not fail. IDK which is preferable for testing purposes.

qxcv reviewed (approved):
High-level: this looks really good! Seems like a win for usability and anti-foot-gunning. Added a few comments, but most are so minor that they can probably be ignored.

ejnnr commented:
> I don't think so. We only normalize at the input layer in networks.build_mlp. It'd actually be a pretty easy change to also add a normalization at the output layer, so perhaps I should just do that now? But it'd be a bit fiddly ensuring that VecNormalize gets removed in the right place, so I lean towards deferring to another PR.

Oh, you're right, I mixed things up in my head. Agreed that we can defer this to a new PR.

ejnnr commented on src/imitation/scripts/common/reward.py:
```diff
@@ -20,12 +21,37 @@ def config():
 
 
+@reward_ingredient.named_config
+def normalize_disable():
```
Assuming we'll want to add output normalization too at some point, we might want to use a different name for these named configs right away. E.g. `normalize_obs_disable` and `normalize_obs_running`, then we can later have similar options for `normalize_rew`. Could also rename `normalize_layer` to e.g. `normalize_input_layer` (since we'll later need `normalize_output_layer`).

OTOH, if we don't end up adding reward normalization for some reason, keeping the current names seems fine. So if you're very uncertain about that, could also change this in another PR

AdamGleave replied:
Good suggestion, changed.

AdamGleave commented on src/imitation/policies/base.py:
```diff
@@ -65,3 +69,27 @@ class FeedForward32Policy(policies.ActorCriticPolicy):
         super().__init__(*args, **kwargs, net_arch=[32, 32])
+
+
+class NormalizeFeaturesExtractor(torch_layers.FlattenExtractor):
```
It will. I don't think we want to normalize CNN input anyway (at least adaptively), as images are usually in a fixed range? I could try to auto-detect whether we're using an MLP or CNN, but I'm reluctant to add too much config magic here.

AdamGleave replied:
Not sure how to solve this so leaving for now, but happy to address it in another PR if you have suggestions.

qxcv replied:
No, that's totally reasonable—I don't normally images most of the time either.

AdamGleave commented on src/imitation/util/networks.py:
```diff
@@ -16,14 +44,83 @@ def forward(self, x):
+        self.running_var += th.square(delta) * self.count * batch_count / tot_count
+        self.running_var /= tot_count
+
+        self.count += batch_count
```
Sadly this doesn't work with a scalar (complains about slice(None)). :( Neat trick though. I could do `self.count.fill_(tot_count)` but that seems obscure.
